hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -444,16 +446,30 @@ private void fireLifecycleEvent(LifecycleState state) {
     }
 
     private boolean doConnectToCandidateCluster(CandidateClusterContext context) {
-        Set<Address> triedAddresses = new HashSet<>();
+        Set<Object> triedTargets = new HashSet<>();
         try {
             waitStrategy.reset();
             do {
-                Collection<Address> addresses = getPossibleMemberAddresses(context.getAddressProvider());
-                for (Address address : addresses) {
+                List<Member> memberList = new ArrayList<>(client.getClientClusterService().getMemberList());
+                if (shuffleMemberList) {
+                    Collections.shuffle(memberList);
+                }
+                //try to connect to a member in the member list first
+                for (Member member : memberList) {
                     checkClientActive();
-                    triedAddresses.add(address);
+                    triedTargets.add(member);
```

Comment:
There is a slight behavior change here but I think this is not big a problem. If some of the member addresses are the same as the addresses configured, then we were not trying to connect the same address twice, but we are doing it right now.

To get the same behavior, maybe we could store the member addresses on the `triedTargets`, and while iterating over the configured addresses, check the contents of it to see whether we tried to connect or not.

Reply:
I missed that there is a loop here, the proposed change causes the configured addresses to not tried more than once in case of a failure, that is probably the cause of the test failures